### PR TITLE
[SPARK-38918][SQL][3.2] Nested column pruning should filter out attributes that do not belong to the current relation

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

Backport #36216 to branch-3.2

### Why are the changes needed?

To fix a bug in `SchemaPruning`.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test

### CODE REVIEW

**Comment by viirya** [comment]
@allisonwang-db Can you fix the test failure?

**Comment by allisonwang-db** [comment]
@viirya I tried a few times but can't reproduce this TPCDSV1_4_PlanStabilitySuite test failure locally...

**Comment by viirya** [comment]
@allisonwang-db Maybe you can just re-trigger the CI?

**Comment by viirya** [comment]
Hmm, from the logs, seems there are unmatched plans:

```
2022-05-04T01:30:18.1915772Z ESC[0m[ESC[0mESC[0minfoESC[0m] ESC[0mESC[0mESC[31m- check simplified (tpcds-v1.4/q5) *** FAILED *** (361 milliseconds)ESC[0mESC[0m
2022-05-04T01:30:18.1916346Z ESC[0m[ESC[0mESC[0minfoESC[0m] ESC[0mESC[0mESC[31m  Plans did not match:ESC[0mESC[0m
2022-05-04T01:30:18.1917069Z ESC[0m[ESC[0mESC[0minfoESC[0m] ESC[0mESC[0mESC[31m  last approved simplified plan: /home/runner/work/spark/spark/sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q5/simplified.txtESC[0mESC[0m
2022-05-04T01:30:18.1918103Z ESC[0m[ESC[0mESC[0minfoESC[0m] ESC[0mESC[0mESC[31m  last approved explain plan: /home/runner/work/spark/spark/sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q5/explain.txtESC[0mESC[0m
```

**Comment by viirya** [comment]
@allisonwang-db I can reproduce it locally by running `build/sbt "sql/testOnly *PlanStabilitySuite -- -z (tpcds-v1.4/q5)"`.

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q4/explain.txt [hunk 1]
```diff
@@ -362,267 +362,267 @@ Output [11]: [customer_id#25, year_total#26, customer_id#50, customer_first_name
 Input [12]: [customer_id#25, year_total#26, customer_id#50, customer_first_name#51, customer_last_name#52, customer_preferred_cust_flag#53, customer_birth_country#54, customer_login#55, customer_email_address#56, year_total#57, customer_id#84, year_total#85]
 
 (54) Scan parquet default.customer
-Output [8]: [c_customer_sk#87, c_customer_id#88, c_first_name#89, c_last_name#90, c_preferred_cust_flag#91, c_birth_country#92, c_login#93, c_email_address#58]
+Output [8]: [c_customer_sk#87, c_customer_id#88, c_first_name#89, c_last_name#90, c_preferred_cust_flag#91, c_birth_country#92, c_login#93, c_email_address#94]
```
Comment (viirya) [inline sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q4/explain.txt:365]: Seems cosmetic change only in the explain string?

**Review by viirya — APPROVED** [review]
lgtm if CI can pass.

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q4/explain.txt (hunk 1 above)
Comment (allisonwang-db) [reply sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q4/explain.txt:365]: Yes and there is no plan change

**Comment by viirya** [comment]
CI still doesn't pass:

```
check simplified (tpcds-v1.4/q4) *** FAILED *** 
```

**Comment by allisonwang-db** [comment]
Hmm let me try again

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @allisonwang-db . Could you rebase this PR once more, please?

**Comment by allisonwang-db** [comment]
@dongjoon-hyun Looks like it still has the same issue:
```bash
# Test passed
build/sbt "sql/testOnly *PlanStabilitySuite"

# Test failed for q4 and q5
build/sbt "sql/testOnly *PlanStabilitySuite -- -z (tpcds-v1.4/q5)"
```

**Comment by dongjoon-hyun** [comment]
Thank you for rechecking.

**Comment by olaky** [comment]
I am facing the same issues here: https://github.com/apache/spark/pull/36753

**Comment by viirya** [comment]
Can #36828 help on stablizing `PlanStabilitySuite`?

**Comment by dongjoon-hyun** [comment]
+1 for @viirya 's comment. @cloud-fan found the root cause and is fixing now on that PR.

**Comment by viirya** [comment]
@allisonwang-db https://github.com/apache/spark/pull/36828 is merged, can you rebase to trigger CI? Thanks.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Review by viirya — APPROVED** [review]
Pending CI

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you so much, @allisonwang-db , @viirya .
